modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -274,192 +210,385 @@ private QueryPlan prepareFragment(String jsonFragment) {
 
-        if (qry != null && qry.state() != QueryState.CLOSED) {
-            assert qry instanceof RootQuery : "Unexpected query object: " + qry;
+                        List<Object> res = new ArrayList<>(rowSize);
 
-            Exception e = new RemoteException(nodeId, msg.queryId(), msg.fragmentId(), msg.error());
+                        for (int i = 0; i < rowSize; i++) {
+                            res.add(root.context().rowHandler().get(i, row));
+                        }
 
-            ((RootQuery<?>) qry).onError(e);
+                        return res;
+                    }));
         }
-    }
 
-    private void onNodeLeft(ClusterNode node) {
-        queryRegistry.runningQueries().forEach(query -> ((Query<?>) query).onNodeLeft(node.id()));
+        private CompletableFuture<Void> close(boolean cancel) {
+            if (!cancelled.compareAndSet(false, true)) {
+                return cancelFut;
+            }
+
+            CompletableFuture<Void> start = new CompletableFuture<>();
+
+            CompletableFuture<Void> end = start
+                    .thenCompose(tmp -> {
+                        if (!root.completeExceptionally(new ExecutionCancelledException())) {
+                            return root.thenAccept(RootNode::close);
+                        }
+
+                        return CompletableFuture.completedFuture(null);
+                    })
+                    .thenCompose(tmp -> {
+                        Map<String, List<CompletableFuture<?>>> requestsPerNode = new HashMap<>();
+                        for (Map.Entry<RemoteFragmentKey, CompletableFuture<Void>> entry : remoteFragmentInitCompletion.entrySet()) {
+                            requestsPerNode.computeIfAbsent(entry.getKey().nodeId(), key -> new ArrayList<>()).add(entry.getValue());
+                        }
+
+                        List<CompletableFuture<?>> cancelFuts = new ArrayList<>();
+                        for (Map.Entry<String, List<CompletableFuture<?>>> entry : requestsPerNode.entrySet()) {
+                            String nodeId = entry.getKey();
+
+                            if (!exchangeSrvc.alive(nodeId)) {
+                                continue;
+                            }
+
+                            cancelFuts.add(
+                                    CompletableFuture.allOf(entry.getValue().toArray(new CompletableFuture[0]))
+                                            .thenRun(() -> {
+                                                try {
+                                                    exchangeSrvc.closeQuery(nodeId, ctx.queryId());
+                                                } catch (IgniteInternalCheckedException e) {
+                                                    throw new IgniteInternalException(
+                                                            "Failed to send cancel message. [nodeId=" + nodeId + ']', e);
```

Comment:
Is it ok the thread that will compose Completable future will fail here due to all 'entry.getValue()' futures can be completed  concurrently?

Reply:
Should `exchangeSrvc.closeQuery` be called if one of `entry.getValue()` will failed?